### ISSUE TITLE
Swagger-ui-react oauth2 redirect url

### ISSUE THREAD

**Issue by thomas-lucas1985** [swagger-api/swagger-ui#5384]
### Q&A (please complete the following information)
 - OS: [Windows 10]
 - Browser: [chrome, safari]
 - Method of installation: [npm]
 - Swagger-UI version: [swagger-ui-react 3.22.2]
 - Swagger/OpenAPI version: [Swagger 2.0]

### How can we help?
I'm aware that the current documentation states that OAuth redirection is not supported. However we are willing to implement 'Try it out' in our react application and are forced to switch to swagger-ui-react to get the 'required parameter issue' solved. In our case, the application is not hosted on the root of the server but in a sub folder. By default when performing authorization, the request includes a redirect uri= hostname/oauth2-redirect.html. In our case this leads to a not found response since it's actually located in a sub folder. It's also not possible to relocate this redirect.html file. 
Is there any way how I can overcome this issue? This is the only thing blocking us at the moment.
If no, would there be a future development to include the support for adjusting the redirect uri?

Thanks in advance

**Comment by Sekramo13** [comment]
This is still an issue on the latest version to date. I'm very confused about why this is still an issue, I see constant issue threads about resolving the issue and they have (mostly) all been closed.

The situation, I want to serve swagger UI from https://localhost:3000/api
After cloning the repo I do the following:

`npm run build`
`docker build -t swag:0.0.26 .`
docker run -p 3000:8080 -e BASE_URL=/api swag:0.0.26

My swagger.json contains basically the default pets.json with the following change:

    "securityDefinitions":{
        "petstore_auth":{
            "type":"oauth2",
            "authorizationUrl":"<DNS>/auth/realms/customer-radar/protocol/openid-connect/auth",
            "flow":"implicit"
        }

Additionally in my index.html in dist/:

    window.onload = function() {
      // Begin Swagger UI call region
      console.error("!!!! Starting Up 01\n\n\n\n")
      print("!!!! Starting Up 02\n\n\n\n")
      
      const ui = SwaggerUIBundle({
        url: "<PathtoMyPublicAccessableSwagger.json>",
        oauth2RedirectUrl: "/tester/oauth2-redirect.html",
        dom_id: '#swagger-ui',
        deepLinking: true,
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
          SwaggerUIBundle.plugins.DownloadUrl
        ],
        layout: "StandaloneLayout"
      })
      // End Swagger UI call region

      window.ui = ui

      ui.initOAuth({
        clientId: "seldon",
        realm: "customer-radar"
      })

    }

If I then hit https://localhost:3000/api I get redirected to swagger-ui.
If I hit authorize I see my client-id is populated with seldon.
If I hit authorize in the pop-up I get redirected to my Idp, I can enter my name and password and then get redirected back:
http://localhost:3039/oauth2-redirect.html#state=%3D&session_state=&access_token=&token_type=bearer&expires_in=900
This redirect fails and I can see from the network traffic monitoring that swagger is posting the redirect URI as part of the request

What I expected to happen:
In swagger-ui/docker/run.sh:
BASE_URL (which is set as part of docker) should change nginx to server all traffic with the new base url.

What is happening:
The redirect that is requested as part of the post to the the Idp does not go through Nginx thus the path will not be updated.

Temporary solution:
In swagger-ui/src/core/index.js this line:
oauth2RedirectUrl: ${window.location.protocol}//${window.location.host}/oauth2-redirect.html,
just update with the desired path
oauth2RedirectUrl: ${window.location.protocol}//${window.location.host}/api/oauth2-redirect.html,

Now it will redirect to my desired base URL of "API"

Additionally you will notice I did set oauth2RedirectUrl: "/tester/oauth2-redirect.html", in index.html.
As far as I can tell it doesn't change anything.

I'm thinking it's not a file/overwrite issue as the oauthinit in the same file is working.

Just reiterating, there is about ~10 open issues that is dealing with this type of issue, there are also numerous closed issues which I'm assuming means someone tried to fix this? (Or do some issues get closed if they go stale, I'm not sure what the policy here is?)

Anyway I'm hoping the above is informative enough to help everyone else with this issue.

**Comment by melone** [comment]
Had the same issue, eventually I solved it using javascript injection with the following env variables as config for the docker container:

``` 
BASE_URL=/api-docs/
OAUTH2_REDIRECT_URL=h" + window.location.protocol.substring(1) + "//" + window.location.host + "/api-docs/oauth2-redirect.html" + "
```

**Comment by mitasov-ra** [comment]
I've found #6591 issue in which using `OAUTH_REDIRECT_URL` considered as a solution, but I think this is NOT a solution.

Just hardcoding app URL into configuration doesn't work at all, especially in containers. I have my swagger-ui container deployed to cloud, and at the moment it has multiple entry points with different hosts and even paths.

For example I have three hosts:
 
 - pretty-host.company.com - this is public host, which maps to UI, API and Swagger UI
 - web-host.namespace.cluster.k8s.company.com - private host, same as first
 - api-v2-host.namespace.cluster.k8s.company.com - private host, just API and Swagger UI

Each of them maps to the same Deployment in Kubernetes.
Swagger UI can be accessed on different paths, respectively:

- /api/v2/docs/ui/
- /api/v2/docs/ui/
- /

In this situation, as you can see, setting `OAUTH_REDIRECT_URL` can make it work only on one of three addresses, which is BAD.

What i've expected is that redirect URL had formed dynamically. Even if it had formed using BASE_URL env variable, such behavior would be more appropriate (even tho I'd had to change base path on the third host), than just hardcoding whole URL into a variable before container is started.

Ideal behavior is forming redirect URL from current URL in browser, because `index.html` and `oauth-redirect.html` lands on the same level and will never have different path prefixes.

P.S. @melone thanks for pointing out there's a possible solution for dynamic host in redirect URL
P.P.S. I'll try to solve that problem, if I succeed I'll make a PR. But I can promise nothing, any help is appreciated

**Comment by mitasov-ra** [comment]
Well, using same approach as @melone, I've come up with this

```js
window.location.href.replace(/index\.html$/, "") + "oauth2-redirect.html"
```

It just works. I dunno how bad this solution is, maybe very bad? :D

**Comment by mitasov-ra** [comment]
My final working solution (from docker-compose):

```yaml
OAUTH2_REDIRECT_URL: '" + window.location.protocol + "//" + window.location.host + window.location.pathname.replace(/index\.html$$/, "") + "oauth2-redirect.html" + "'
```

**Comment by mitasov-ra** [comment]
This issue is fixed in #7476. Now `oauth2RedirectUrl` includes `window.location.pathname` by default and @melone's and mine workarounds are not necessary anymore.
